Schema details

AlexAxthelm opened the pull request:
multi-issue PR to address open details for JSON schema

Builds off #251 
Closes #237 (types)
Closes #238 (regex & length)
Closes #239 (enums)
Closes #241 (required)
Closes #240 (descriptions)

AlexAxthelm committed:
Remove pathway_type_tooltip

AlexAxthelm committed:
refactor(schema): match field types to data model

Match types for fields against defined data model.

Closes #237

AlexAxthelm committed:
refactor(schema): update `sectors` schema

includes simplifying the `technologies` validation, by using a simple
set of technologies, rather than sector-specific lists

AlexAxthelm committed:
refactor(schema): JSON Schema validation constraints

* handle string length limits and regex, closes #238
* Check enum options, closes #239

AlexAxthelm committed:
refactor(schema): harmonize required fields with model

Closes #241

AlexAxthelm committed:
JS escaping for regex

AlexAxthelm committed:
Update example files to validate against schema

AlexAxthelm committed:
`prettier`

AlexAxthelm committed:
docs(schema): Update JSON schema field descriptions.

Closes #240

AlexAxthelm commented on pbtar_schema.json:
```diff
@@ -265,9 +202,7 @@
-      "pathway_type",
-      "pathway_type_tooltip",
-      "modelYearEnd",
+      "pathwayType",
```
@jdhoffa @jacobvjk Can you confirm that `modelYearStart` and `modelYearEnd` are not mandatory?

jdhoffa replied:
Agree with Copilot here, `modelYearEnd` is currently used pretty extensively in the website, especially in any "target year" filtering: 
<img width="825" height="683" alt="Screenshot 2025-08-13 at 10 48 58" src="https://github.com/user-attachments/assets/9b82da24-0cdf-44b6-88c9-20cc27850831" />

I leave it mainly to @lloyd-rmi and @jacobvjk on how we want to handle this. We can remove the ability to filter by the target year of the scenario, or we can make the field mandatory. 
(Or, alternatively, I suppose we could make it nullable, but then if anyone filters for a specific Target Year, nothing would appear). 

@AlexAxthelm thoughts?

AlexAxthelm replied:
We should probably add some tests, but my suspicion is that it's fine technically to omit `modelYearStart`, `modelYearEnd`, and `modelYearNetzero`. They just would never show up if someone filters on that field (`Target Year == 2040 ` or `Target Year < 2051`) but should appear on the other filters (`Temperature < 2`).

Overall, I'd say merge as-is, and there will be a second round to address changes once we see this schema in practice.

jdhoffa replied:
I have a call with Jacob and Lloyd to discuss how we want to address this in the long-term (and agree, we don't need to decide that now). 

However, can you please add 2 tests for the cases you just identified? If we want to merge this as-is, then I want to really be sure that it functions as expected in those edge cases. 

e.g. A scenario with no `modelYearEnd` still appears when the `Temperature == 1.5` filter is selected, but does not appear when the `Target Year == 2040` 

Otherwise I fear we will forget this edge-case

AlexAxthelm replied:
I've added a minimal test scenario to the `src/data` directory, so that we can visually inspect it. There's some weirdness visually (empty badge) when theres nothing in the "targets" section on the card, and the filter dropdowns now show "undefined", but that isn't a real filter (same as not setting it). but on the other hand, the scenario is showing up in the filters for which it can be filtered.

TL;DR: I wouldn't say that omitting these fields from a scenario file _breaks_ anything, but if we're going to allow them to be unset, we need to handle unset values better.

I think that this should be a follow-up in a separate ticket to make a decision and implement.

<img width="1014" height="608" alt="Screenshot_2025-08-14_08 36 09" src="https://github.com/user-attachments/assets/130fe90d-bd4a-4c66-96ca-a3b8fec38c16" />

jdhoffa replied:
Thanks for looking into this Alex. 

> I think that this should be a follow-up in a separate ticket to make a decision and implement.

Agree. Lloyd, Jacob and I will discuss this today, and can get a decision, not necessary to block this PR on that decision. @AlexAxthelm would you mind making that separate ticket so that we follow-up?

AlexAxthelm replied:
#255

jdhoffa replied:
Thank you!

AlexAxthelm committed:
Merge branch 'main' into schema-details

jdhoffa reviewed (commented):
Mostly looks good to me, pending resolution of the open comment.

AlexAxthelm committed:
Add minimal validating scenario

AlexAxthelm committed:
test(search): search in the absence of non-mandatory fields

add test that filtering works for scenarios that are missing fields that
are not marked as "required" in the JSON schema

jdhoffa reviewed (approved):
Looks great, thank you!
